### PR TITLE
Add Word2Vec Transforms

### PR DESCRIPTION
**Pull request by JenZhao**
still adding the word2vec steps afterwards

### CODE REVIEW

File: hidi/factorization.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        :rtype: str
+        """
+        if 'item_id' not in df.index:
+            df.set_index('item_id', inplace=True)
+
+        words = ''
+        for index in df.index.unique():
+            a0 = df.loc[index].link_id
+            a = a0
+            for i in range(self.n_shuffles-1):
```
Comment (kahnjw) [inline hidi/factorization.py]: Whitespace around `-`.

File: hidi/factorization.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        :type df: pandas.Dataframe
+        :rtype: str
+        """
+        if 'item_id' not in df.index:
+            df.set_index('item_id', inplace=True)
+
+        words = ''
+        for index in df.index.unique():
+            a0 = df.loc[index].link_id
+            a = a0
```
Comment (kahnjw) [inline hidi/factorization.py]: Can we name these variables with something more than 1-2 letters? Something that describes what they are.

File: hidi/factorization.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            for i in range(self.n_shuffles-1):
+                a = np.append(a, permutation(a0))
+            b = " ".join(str(x) for x in a)
```
Comment (kahnjw) [inline hidi/factorization.py]: For consistency, lets use single quotes `'` everywhere except docstrings.

File: hidi/factorization.py (hunk 2 above)
Comment (JenZhao) [inline hidi/factorization.py]: will do

File: hidi/factorization.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            words = " ".join([words, b]).strip()
+        return words, kwargs
```
Comment (kahnjw) [inline hidi/factorization.py]: We need to add the option of sending an `order` column so the words can be explicitly ordered. @jrecursive may have more specifics on this.

Comment (JenZhao) [inline hidi/factorization.py]: order for the item_id?

File: hidi/factorization.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            a0 = df.loc[index].link_id
+            a = a0
+            for i in range(self.n_shuffles - 1):
+                a = np.append(a, permutation(a0))
+            b = ' '.join(str(x) for x in a)
+            words = ' '.join([words, b]).strip()
+        return words, kwargs
+
+
+class W2VGenismTransform(Transform):
```
Comment (kahnjw) [inline hidi/factorization.py]: `Genism` -> `Gensim`

File: hidi/factorization.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                                         permutation(temp_item_id_list))
+            joined_shuffled_list = ' '.join(str(x) for x in item_id_list)
+            words = ' '.join([words, joined_shuffled_list]).strip()
+        return words, kwargs
+
+
+class W2VGenismTransform(Transform):
+    """
+    Generalized transform for gensim.models.Word2Vec
+    Takes an uninitialized gensim.models.Word2Vec and here is more details
```
Comment (kahnjw) [inline hidi/factorization.py]:
```
Takes an uninitialized :code:`gensim.models.Word2Vec`. Read more about this model:

<link>
```

File: hidi/factorization.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            `reverse_dictionary` `data` is the tokenized words, `count` is a
+            list of tuple which consists of `(item, count)`, `dictionary`
+            stores tokens of each items as its keys and items as its values
+            and `reverse_dictionary` is the reversed of `dictionary`
+        """
+        if isinstance(words, str):
+            words = words.split()
+        count = [['UNK', -1]]
+        count_words = collections.Counter(words)
+        count.extend(count_words.most_common((self.vocabulary_size-1)))
```
Comment (kahnjw) [inline hidi/factorization.py]: Not sure why flake8 isnt catching this. Add some whitespace around `-`:

```python
count.extend(count_words.most_common((self.vocabulary_size - 1)))
```

File: hidi/factorization.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        self.n_shuffles = n_shuffles
+
+    def transform(self, df_temp, **kwargs):
+        """
+        :param df: a pandas Dataframe with two columns:
+        :code:`link_id`, :code:`item_id`
+        :type df: pandas.Dataframe
+        :rtype: a string of shuffled item_id
+        """
+        df = df_temp
```
Comment (kahnjw) [inline hidi/factorization.py]: This looks suspicious. Is the intent here to copy `df_temp`? If so, I think this only copies the reference to the underlying dataframe object. To actually copy it, one needs to do:

```python
df  = df_temp.copy(deep=deep)
```

Where deep is a boolean.

File: hidi/factorization.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        words = ''
+        for index in df.index.unique():
+            temp_item_id_list = df.loc[index].item_id
+            item_id_list = temp_item_id_list
+            for i in range(self.n_shuffles - 1):
+                item_id_list = np.append(item_id_list,
+                                         permutation(temp_item_id_list))
+            joined_shuffled_list = ' '.join(str(x) for x in item_id_list)
+            words = ' '.join([words, joined_shuffled_list]).strip()
+        return words, kwargs
```
Comment (kahnjw) [inline hidi/factorization.py]: At some point, we should see how long this takes, and benchmark it against `StringIO`. I have no idea which would be faster, but I think both should be possible.

**Comment by kahnjw** [comment]
@JenZhao should we merge this?

**Comment by JenZhao** [comment]
going to make a few more changes this week.

**Pull request merged**